bigtable/client/filters.h

```diff
@@ -0,0 +1,308 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_FILTERS_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_FILTERS_H_
+
+#include <bigtable/client/version.h>
+#include <google/bigtable/v2/data.pb.h>
+#include <chrono>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Define the interfaces to create filter expressions.
+ *
+ * Example:
+ * @code
+ * // Get only data from the "fam" column family, and only the latest value.
+ * auto filter = Filter::Chain(Filter::Family("fam"), Filter::Latest(1));
+ * table->ReadRow("foo", std::move(filter));
+ * @endcode
+ *
+ * Those filters that use regular expressions, expect the patterns to be in
+ * the [RE2](https://github.com/google/re2/wiki/Syntax) syntax.
+ *
+ * @note Special care need be used with the expression used. Some of the
+ *   byte sequences matched (e.g. row keys, or values), can contain arbitrary
+ *   bytes, the `\C` escape sequence must be used if a true wildcard is
+ *   desired. The `.` character will not match the new line character `\n`,
+ *   which may be present in a binary value.
+ */
+class Filter {
+ public:
+  /// An empty filter, discards all data.
+  Filter() : filter_() {}
+
+  // TODO() - replace with = default if protobuf gets move constructors.
+  Filter(Filter&& rhs) noexcept : Filter() { filter_.Swap(&rhs.filter_); }
+
+  // TODO() - replace with = default if protobuf gets move constructors.
+  Filter& operator=(Filter&& rhs) noexcept {
+    Filter tmp(std::move(rhs));
+    tmp.filter_.Swap(&filter_);
+    return *this;
+  }
+
+  Filter(Filter const& rhs) = default;
+  Filter& operator=(Filter const& rhs) = default;
+
+  /// Create a filter that accepts only the last @a n values.
+  static Filter Latest(int n) {
+    Filter result;
+    result.filter_.set_cells_per_column_limit_filter(n);
+    return result;
+  }
+
+  /**
+   * Return a filter that matches column families matching the given regexp.
+   *
+   * @param pattern the regular expression.  It must be a valid
+   *     [RE2](https://github.com/google/re2/wiki/Syntax) pattern.
+   *     For technical reasons, the regex must not contain the ':' character,
+   *     even if it is not being used as a literal.
+   */
+  static Filter Family(std::string pattern) {
```

Comment:
Can we name the filters that take regex as `FamilyRegex`, `ColumnRegex`, etc?

Reply:
Done.  If I miss any please let me know.